byzcoin/service.go

```diff
@@ -1553,6 +1557,43 @@ func newService(c *onet.Context) (onet.Service, error) {
 		return nil, err
 	}
 
+	ver, err := s.LoadVersion()
+	if err != nil {
+		return nil, err
+	}
+	switch ver {
+	case 0:
+		// Version 0 means it hasn't been set yet. If there are any ByzCoin_[0-9af]+
+		// buckets, then they must be old format.
+		db, _ := s.GetAdditionalBucket([]byte("check-db-version"))
+
+		// Look for a bucket that has a byzcoin database in it.
+		err := db.View(func(tx *bolt.Tx) error {
+			c := tx.Cursor()
+			for k, _ := c.First(); k != nil; k, _ = c.Next() {
+				log.Lvlf4("looking for old ByzCoin data in bucket %v", string(k))
+				if existingDB.Match(k) {
```

Comment:
My understanding is that we're looping over bucket names. But here's you're checking if the bucket name matches `^ByzCoin_[0-9af]+`. But isn't that true for the new database too? In `createStateTrie` we name buckets based on the hex skipchain ID.

Reply:
also, why isn't it `[0-9a-f]`?

Reply:
On the first start of a db that will be new-style, it has no buckets that match ByzCoin_[0-9a-f]+ yet. So we get to the end of the loop, set the version to 1, and never do the scan again. On an attempted start with a database that has been used with the collections code, we match the first one, throw the error, and refuse to start.